File: src/negative_test.py
"""Negative testing engine to prove that high CVSS != high priority."""

from typing import List, Optional

from src.models import (
    NegativeTestItem,
    OrganizationProfile,
    PriorityLevel,
    TriageResult,
    Vulnerability,
)
from src.ranking import rank_all_vulnerabilities


def explain_negative_test_result(
    triage_item: TriageResult,
    profile: OrganizationProfile,
    total_candidates: int,
) -> str:
    """Generate a step-by-step diagnostic explaining why a high-CVSS CVE ranked low or was de-prioritized."""
    breakdown = triage_item.score_breakdown
    vuln = triage_item.vulnerability
    reasons: List[str] = []

    # 1. Check KEV
    if not vuln.cisa_kev:
        reasons.append(
            f"Zero exploitation evidence in CISA KEV (0.0 pts awarded out of {profile.weight_modifiers.cisa_kev_weight * 100:.0f} max KEV points)."
        )

    # 2. Check EPSS
    epss = vuln.first_epss or 0.0
    if epss < 0.05:
        reasons.append(
            f"Extremely low 30-day exploitation probability (EPSS {epss:.2%} contributes only {breakdown.epss_contribution:.1f} pts)."
        )
    elif epss < 0.20:
        reasons.append(
            f"Low exploitation probability (EPSS {epss:.2%} contributes only {breakdown.epss_contribution:.1f} pts)."
        )

    # 3. Check Critical Product Context
    if not breakdown.is_critical_product:
        reasons.append(
            f"Asset '{vuln.product_name}' is NOT in {profile.name}'s critical products list ({', '.join(profile.critical_products)}), forfeiting the 1.4x priority multiplier."
        )

    # 4. Check Profile Weighting Impact
    cvss_wt = profile.weight_modifiers.cvss_weight
    if cvss_wt <= 0.35:
        reasons.append(
            f"{profile.name}'s risk appetite allocates only {cvss_wt:.0%} weight to theoretical severity (CVSS), prioritizing real-world exploitation signals instead."
        )

    explanation = (
        f"Despite a high technical severity of CVSS {vuln.cvss_base_score:.1f}, {vuln.cve_id} ranked #{triage_item.rank} "
        f"of {total_candidates} candidates (Score: {breakdown.final_score:.1f}/100, Priority: {triage_item.priority.value}). "
        f"Key de-prioritization factors: " + " ".join(reasons)
    )
    return explanation


def find_negative_test_candidates(
    vulnerabilities: List[Vulnerability],
    profile: OrganizationProfile,
    min_cvss: float = 9.0,
    max_rank_threshold: int = 15,
) -> List[NegativeTestItem]:
    """Find vulnerabilities with CVSS >= min_cvss that ranked outside the top tier for the given organization."""
    all_ranked = rank_all_vulnerabilities(vulnerabilities, profile)
    total_candidates = len(all_ranked)
    candidates: List[NegativeTestItem] = []

    for item in all_ranked:
        cvss = item.vulnerability.cvss_base_score or 0.0
        if cvss >= min_cvss and (item.rank > max_rank_threshold or item.priority in (PriorityLevel.LOW, PriorityLevel.MEDIUM)):
            reason = explain_negative_test_result(item, profile, total_candidates)
            candidates.append(
                NegativeTestItem(
                    vulnerability=item.vulnerability,
                    score_breakdown=item.score_breakdown,
                    rank=item.rank,
                    total_candidates=total_candidates,
                    reason_low_or_excluded=reason,
                )
            )

    return candidates

File: vulnlens_mobile/lib/services/negative_test.dart
import '../models/models.dart';
import 'ranking.dart';

String explainNegativeTestResult(
  TriageResult triageItem,
  OrganizationProfile profile,
  int totalCandidates,
) {
  final breakdown = triageItem.scoreBreakdown;
  final vuln = triageItem.vulnerability;
  final reasons = <String>[];

  // 1. Check KEV
  if (!vuln.cisaKev) {
    final maxKevPts = (profile.weightModifiers.cisaKevWeight * 100).toInt();
    reasons.add(
        'Zero exploitation evidence in CISA KEV (0.0 pts awarded out of $maxKevPts max KEV points).');
  }

  // 2. Check EPSS
  final epss = vuln.firstEpss ?? 0.0;
  if (epss < 0.05) {
    final epssPct = (epss * 100).toStringAsFixed(2);
    reasons.add(
        'Extremely low 30-day exploitation probability (EPSS $epssPct% contributes only ${breakdown.epssContribution.toStringAsFixed(1)} pts).');
  } else if (epss < 0.20) {
    final epssPct = (epss * 100).toStringAsFixed(2);
    reasons.add(
        'Low exploitation probability (EPSS $epssPct% contributes only ${breakdown.epssContribution.toStringAsFixed(1)} pts).');
  }

  // 3. Check Critical Product Context
  if (!breakdown.isCriticalProduct) {
    final critList = profile.criticalProducts.join(', ');
    reasons.add(
        "Asset '${vuln.productName}' is NOT in ${profile.name}'s critical products list ($critList), forfeiting the 1.4x priority multiplier.");
  }

  // 4. Check Profile Weighting Impact
  final cvssWt = profile.weightModifiers.cvssWeight;
  if (cvssWt <= 0.35) {
    final cvssPct = (cvssWt * 100).toInt();
    reasons.add(
        "${profile.name}'s risk appetite allocates only $cvssPct% weight to theoretical severity (CVSS), prioritizing real-world exploitation signals instead.");
  }

  final cvssStr = vuln.cvssBaseScore?.toStringAsFixed(1) ?? 'N/A';
  final scoreStr = breakdown.finalScore.toStringAsFixed(1);
  return 'Despite a high technical severity of CVSS $cvssStr, ${vuln.cveId} ranked #${triageItem.rank} of $totalCandidates candidates (Score: $scoreStr/100, Priority: ${triageItem.priority.label}). Key de-prioritization factors: ${reasons.join(' ')}';
}

List<NegativeTestItem> findNegativeTestCandidates(
  List<Vulnerability> vulnerabilities,
  OrganizationProfile profile, {
  double minCvss = 9.0,
  int maxRankThreshold = 10,
}) {
  final allRanked = rankAllVulnerabilities(vulnerabilities, profile);
  final totalCandidates = allRanked.length;
  final candidates = <NegativeTestItem>[];

  for (final item in allRanked) {
    final cvss = item.vulnerability.cvssBaseScore ?? 0.0;
    if (cvss >= minCvss &&
        (item.rank > maxRankThreshold ||
            item.priority == PriorityLevel.low ||
            item.priority == PriorityLevel.medium)) {
      final reason = explainNegativeTestResult(item, profile, totalCandidates);
      candidates.add(
        NegativeTestItem(
          vulnerability: item.vulnerability,
          scoreBreakdown: item.scoreBreakdown,
          rank: item.rank,
          totalCandidates: totalCandidates,
          reasonLowOrExcluded: reason,
        ),
      );
    }
  }

  return candidates;
}
